src/OrchardCore/OrchardCore.Data/Documents/FileSystemDocumentStore.cs

```diff
@@ -0,0 +1,88 @@
+using System;
+using System.IO;
+using System.Reflection;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Locking;
+
+namespace OrchardCore.Data.Documents
+{
+    /// <summary>
+    /// A scoped <see cref="IFileDocumentStore"/> implementation using
+    /// file system to persist documents.
+    /// </summary>
+    public class FileSystemDocumentStore : FileDocumentStore
+    {
+        private const string LockKey = "FILESYSTEM_DOCUMENTSTORE_LOCK";
```

Comment:
Why is it necessary? There should be a single server accessing the file system, or this should be using the database.

Reply:
It's to prevent concurrent access on the same file.
Original implementation of `FileDocumentStore` had a single SemaphoreSlim to achieve it, here we have one lock per file.

Reply:
> There should be a single server accessing the file system
> It's to prevent concurrent access on the same file.

What are the concurrent things accessing the same file?

Reply:
Two concurrent requests ?

Reply:
That's my point, this is what we expect, and that's why it was using a semaphore which is much more reliable than creating temporary files which will be prone to the same concurrent issues it would try to fix. Rephrasing: if two threads can access files in read/write and this can be an issue, we shouldn't solve it by using a file in read/write ...

Reply:
Sorry I don't understand the problem. It uses `ILocalLock` which is a per tenant dictionary of Semaphores.

Reply:
Sorry I missread the code then, I thought "FILESYSTEM_DOCUMENTSTORE_LOCK" was a filename.